WIP Subscription rework

brianshannan opened the pull request:
Changes:
- Return subscriptions from `subscribe()` instead of a subscription id
- Allow iterating over messages in subscription with `async for`. To use, don't specify a callback when subscribing
- Move `unsubscribe()` and `drain(sid=sid)` to the subscription class
- Remove `publish_request()`, `publish()` now has a `reply` kwarg
- Remove `auto_unsubscribe()`, `subscribe()` already had a kwarg for this
- Remove support for executing synchronous callbacks
  - Remove `subscribe_async()`
  - Remove `is_async` kwarg from `subscribe()`
- Remove support for executing callbacks as tasks
- Remove multiple response functionality from `request()`
- Remove `timed_request()`, name is very misleading as all requests are timed. This was really old style requests, moved to a kwarg of `request()` called `old_style`
- Made connection context manager async
- Various internal code reoganizations
- Various doc cleanups
- Cleaned up example code

charliestrawn commented on examples/nats-req/__main__.py:
```diff
@@ -72,15 +75,25 @@ async def reconnected_cb():
-    print("Received a message on '{subject} {reply}': {data}".format(
-        subject=subject, reply=reply, data=data))
+    print(
+        "Received a message on '{subject} {reply}': {data}".format(
```
nit: f-string

brianshannan replied:
I do agree fstrings would be good, I'm going to tackle those in a followup pr as this one is going to be large enough as is

charliestrawn commented on tests/client_async_await_test.py:
```diff
@@ -21,7 +21,8 @@ async def subscription_handler(msg):
             msgs.append(msg)
```
nit: rename to `test_client_async_await.py`

charliestrawn commented on tests/client_test.py:
```diff
@@ -231,7 +231,7 @@ async def subscription_handler(msg):
 
```
nit: rename `test_client.py`

charliestrawn commented on tests/client_test.py:
```diff
@@ -1841,39 +1770,42 @@ async def closed_cb():
+            await nc2.publish(
+                "foo",
+                b'help',
+                reply="my-replies.%s" % nc._nuid.next().decode()
```
nit: couple of old style formatting, suggest f-strings

charliestrawn commented on nats/aio/client.py:
```diff
@@ -961,33 +904,23 @@ async def timed_request(self, subject, payload, timeout=0.5):
         if limit > 0:
             b_limit = ("%d" % limit).encode()
-        b_sid = ("%d" % sid).encode()
+        b_sid = ("%d" % sub._id).encode()
```
nit: old style formatting

charliestrawn commented on nats/aio/client.py:
```diff
@@ -1437,38 +1370,38 @@ async def _process_msg(self, sid, subject, reply, data):
         # Check if it is an old style request.
-        if sub.future is not None:
-            if sub.future.cancelled():
+        if sub._future:
```
arbitrary line for discussion: pre-existing, but it seems like a potential smell that client is accessing a lot of subscriptions internals. Sub being a parameter to a lot of these functions also makes me think that some of the behavior could be moved to the Subscription class. That may help isolate future changes and hide some of the complexity of these conditionals, making this class a little cleaner.

brianshannan replied:
Agreed, I got a fair amount of this code moved

wallyqs commented on nats/aio/client.py:
```diff
@@ -72,29 +72,169 @@
+        if self._conn.is_draining:
+            raise ErrConnectionDraining
+
+        self._stop_processing()
```
I think here we need a special case for when `limit` is 0 so that it can replace `auto_unsubscribe` usage like:

```python
    # Stop receiving after 2 messages.
    sub = await nc.subscribe("foo", cb=message_handler)
    await sub.unsubscribe(limit=2)
    await nc.publish("foo", b'Hello')
    await nc.publish("foo", b'World')
    await nc.publish("foo", b'Not delivered')
    await nc.flush()
```

Then the 3rd message would not be delivered.

```suggestion
        if limit == 0:
                self._stop_processing()
```

wallyqs commented on nats/aio/client.py:
```diff
@@ -1437,38 +1341,38 @@ async def _process_msg(self, sid, subject, reply, data):
-        sub.received += 1
-        if sub.max_msgs > 0 and sub.received >= sub.max_msgs:
+        sub._received += 1
+        if sub._max_msgs > 0 and sub._received >= sub._max_msgs:
```
I might be wrong but maybe should call `sub._stop_processing` here to mark the `_wait_for_msgs_task` for the subscriptions as done?

wallyqs commented on tests/client_test.py:
```diff
@@ -420,30 +395,35 @@ async def subscription_handler(msg):
 
         await nc.connect()
-        sid = await nc.subscribe("tests.>", cb=subscription_handler)
+        sub = await nc.subscribe('tests.>', iterate=True)
```
How about if `cb` being `None` would mean that the subscription has to be used in the iterator style? That would feel slightly more similar to how the [nats.rs/Rust client does it](https://github.com/nats-io/nats.rs/blob/b6e04c4768a0580634dfb03c4ca252b2ebf245f1/examples/nats-box/main.rs#L73), and also would not need to have `subscribe_sync` like in the Go client.  Wdyt?

```python
sub = await nc.subscribe('tests.>')
...
async for msg in sub.messages:
...
```

brianshannan replied:
Yeah I'm fine with that. I need to edit the docs still, I'll add something that explicitly says that so consumers are aware.

wallyqs reviewed (commented):
Thanks a lot, this is looking great.  I left some comments for now but will resume later.

brianshannan committed:
make api more consistent with other clients

This commit is an attempt to make the api of this client more closely
match the api of nats clients in other languages. Below is a non
exhaustive list of the changes made.

- Return subscriptions from `subscribe()` instead of a subscription id
- Allow iterating over messages in subscription with `async for`. To use, don't specify a callback when subscribing
- Move `unsubscribe()` and `drain(sid=sid)` to the subscription class
- Remove `publish_request()`, `publish()` now has a `reply` kwarg
- Remove `auto_unsubscribe()`, `subscribe()` already had a kwarg for this
- Remove support for executing synchronous callbacks
  - Remove `subscribe_async()`
  - Remove `is_async` kwarg from `subscribe()`
- Remove support for executing callbacks as tasks
- Remove multiple response functionality from `request()`
- Remove `timed_request()`, name is very misleading as all requests are timed. This was really old style requests, moved to a kwarg of `request()` called `old_style`
- Made connection context manager async
- Various internal code reoganizations
- Various doc cleanups
- Cleaned up example code
